### PR TITLE
📥 Implicit Total Space Conversion

### PR DESCRIPTION
**Pull request by mmcqd**
Hello! I'm a first time cooltt contributor :)

This PR makes implicit the conversion between a signature indexed type family and its total space (as implemented in #280), and replaces `as` with `#` as the patch operator. Now instead of writing `fam # []`, one can write `fam` and `tp # [ patches ]` can be written if `tp` is either a signature indexed type family or a signature. This takes cooltt one step closer to the original total space conversion proposal in #267. 
A patch now looks like this:
```
def el : type := sig (A : type) (a : A)
def el-patch : type := el # [ A .= nat | a .= 4 ]
```
The definition of a category now looks like this:
```
def category : type :=
  sig (ob : type)
      (hom : sig (s : ob) (t : ob) → type)
      (idn : (x : ob) -> hom # [ s .= x | t .= x ])
      (seq : (f : hom) -> (g : hom # [ s .= f.t ]) -> hom # [ s .= f.s | t .= g.t ])
      (seqL : (f : hom) -> path {hom # [ s .= f.s | t .= f.t ]} {seq {idn {f.s}} f} f)
      (seqR : (f : hom) -> path {hom # [ s .= f.s | t .= f.t ]} {seq f {idn {f.t}}} f)
      (seqA : (f : hom) -> (g : hom # [ s .= f.t ]) -> (h : hom # [ s .= g.t ]) -> path {hom # [ s .= f.s | t .= h.t ]} {seq f {seq g h}} {seq {seq f g} h})
```

I may have violated some cooltt style guidelines or misunderstood some implementation details, and would be happy to refactor this as needed. I wasn't sure where to put the new tactic which handles this, so I put it in `Refiner.Structural`, but this may have been the wrong place.

If this goes well, I also have some ideas for making it more convenient to inhabit patched record types that I'd love to try!

### CODE REVIEW

**Comment by favonia** [comment]
Thank you for your PR and we will take a careful look.

Quick response to this question:

> I may have violated some cooltt style

As long as you are (1) removing trailing whitespace, (2) using `ocp-indent` for OCaml indentation, and (3) being semi-consistent with the surrounding code (we do not have global naming convention etc.), then your coding style should be fine!

**Comment by mmcqd** [comment]
> This looks really really cool. Btw what about writing `#` instead of `as`?

I agree that `#` would be nicer as the patch operator. Done.

**Review by TOTBWF — CHANGES_REQUESTED** [review]
Looks great, amazing first contribution :)

A couple of minor nits, mostly related to organization. With these changes, it seems to me that `Tactics.ml` would be a better home for these tactics, as it's more in line with `intro_implicit_connectives` et. al.

File: src/core/Refiner.ml
Comment (mmcqd) [reply src/core/Refiner.ml]: I agree, I've moved it. Is `intro_conversions` a good name for this? Perhaps `intro_implicit_conversions` would be better? I made it plural with the expectation there might be more conversions of this style done in the future.

File: src/core/Refiner.ml
Comment (mmcqd) [reply src/core/Refiner.ml:751]: I agree it makes more sense there, but it depends on `quote_sign_codes`, which is also used by `patch_fields`. Should that be exposed by the refiner?

**Comment by favonia** [comment]
> This looks really really cool. Btw what about writing `#` instead of `as`?

As a side issue that belongs to another PR, do we prefer to drop `#` completely? PS: it is widely accepted that either :genie: can do it or it is mathematically impossible.

**Comment by jonsterling** [comment]
@favonia if we could drop it that would be awesome!

File: src/core/Refiner.ml
Comment (TOTBWF) [reply src/core/Refiner.ml:751]: Ah, good point. Perhaps the move is to expose `quantifiers` somewhere. Either way, this is a minor nit, and shouldn't hold up this PR. Great work!

**Review by TOTBWF — APPROVED** [review]
Looks great! Welcome to the 😎 zone

**Comment by mmcqd** [comment]
> Looks great! Welcome to the 😎 zone

Cool! Looking forward to making some more contributions 👀. btw I can't merge the PR because I don't have write access, so one of you will have to do it

**Pull request merged**
